sqlglot/optimizer/scope.py

```diff
@@ -86,19 +86,32 @@ def _collect(self):
         self._derived_tables = []
         self._raw_columns = []
 
-        for node, *_ in _walk_next_scope(self.expression):
+        # We'll use this variable to pass state into the dfs generator.
+        # Whenever we set it to True, we exclude a subtree from traversal.
+        prune = False
```

Comment:
This is a wacky way to control the DFS traversal... but I kinda dig it.

Reply:
does this reval the lambda????

Reply:
Lambdas can reference variables in the scope they're defined. The values don't get copied. 💥 

So every time the lambda function gets called, it returns the current value of `prune`. And the lambda gets called on every node.